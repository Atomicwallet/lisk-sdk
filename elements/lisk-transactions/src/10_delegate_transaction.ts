/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */

import { BaseTransaction, StateStore } from './base_transaction';
import { CHAIN_STATE_DELEGATE_USERNAMES, DELEGATE_NAME_FEE } from './constants';
import { TransactionError } from './errors';
import { BaseTransactionInput } from './types';

interface RegisteredDelegate {
	readonly username: string;
	readonly address: Buffer;
}
interface ChainUsernames {
	readonly registeredDelegates: RegisteredDelegate[];
}

export interface DelegateAsset {
	readonly username: string;
}

export const delegateRegistrationAssetSchema = {
	$id: 'lisk/delegate-registration-transaction',
	type: 'object',
	required: ['username'],
	properties: {
		username: {
			dataType: 'string',
			fieldNumber: 1,
			minLength: 1,
			maxLength: 20,
		},
	},
};

export class DelegateTransaction extends BaseTransaction {
	public static TYPE = 10;
	public static NAME_FEE = BigInt(DELEGATE_NAME_FEE);
	public static ASSET_SCHEMA = delegateRegistrationAssetSchema;
	public readonly asset: DelegateAsset;

	public constructor(transaction: BaseTransactionInput<DelegateAsset>) {
		super(transaction);

		this.asset = transaction.asset;
	}

	protected async applyAsset(
		store: StateStore,
	): Promise<ReadonlyArray<TransactionError>> {
		const errors: TransactionError[] = [];
		const sender = await store.account.get(this.senderId);

		// Data format for the registered delegates
		// chain:delegateUsernames => { registeredDelegates: { username, address }[] }
		const usernamesBuffer = await store.chain.get(
			CHAIN_STATE_DELEGATE_USERNAMES,
		);
		const usernames = usernamesBuffer
			? (JSON.parse(usernamesBuffer.toString('utf8')) as ChainUsernames)
			: { registeredDelegates: [] };
		const usernameExists = usernames.registeredDelegates.find(
			delegate => delegate.username === this.asset.username,
		);

		if (!usernameExists) {
			usernames.registeredDelegates.push({
				username: this.asset.username,
				address: this.senderId,
			});
			usernames.registeredDelegates.sort((a, b) =>
				a.address.compare(b.address),
			);
			store.chain.set(
				CHAIN_STATE_DELEGATE_USERNAMES,
				Buffer.from(JSON.stringify(usernames), 'utf8'),
			);
		}

		if (usernameExists) {
			errors.push(
				new TransactionError(
					`Username is not unique.`,
					this.id,
					'.asset.username',
				),
			);
		}
		if (sender.isDelegate || sender.username) {
			errors.push(
				new TransactionError(
					'Account is already a delegate',
					this.id,
					'.asset.username',
				),
			);
		}
		sender.username = this.asset.username;
		sender.isDelegate = 1;
		store.account.set(sender.address, sender);

		return errors;
	}

	protected async undoAsset(
		store: StateStore,
	): Promise<ReadonlyArray<TransactionError>> {
		const sender = await store.account.get(this.senderId);

		// Data format for the registered delegates
		// chain:delegateUsernames => { registeredDelegates: { username, address }[] }
		const usernamesBuffer = await store.chain.get(
			CHAIN_STATE_DELEGATE_USERNAMES,
		);
		const usernames = usernamesBuffer
			? (JSON.parse(usernamesBuffer.toString('utf8')) as ChainUsernames)
			: { registeredDelegates: [] };
		const updatedRegisteredDelegates = {
			registeredDelegates: usernames.registeredDelegates.filter(
				delegate => delegate.username !== sender.username,
			),
		};
		updatedRegisteredDelegates.registeredDelegates.sort((a, b) =>
			a.address.compare(b.address),
		);
		store.chain.set(
			CHAIN_STATE_DELEGATE_USERNAMES,
			Buffer.from(JSON.stringify(updatedRegisteredDelegates), 'utf8'),
		);

		sender.username = null;
		sender.isDelegate = 0;
		store.account.set(sender.address, sender);
		return [];
	}
}
