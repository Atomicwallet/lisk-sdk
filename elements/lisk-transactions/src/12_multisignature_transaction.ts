/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */

import {
	getAddressAndPublicKeyFromPassphrase,
	getAddressFromPublicKey,
	signData,
	bufferToHex,
	hash,
} from '@liskhq/lisk-cryptography';

import { BaseTransaction, StateStore } from './base_transaction';
import { TransactionError } from './errors';
import { createResponse, TransactionResponse } from './response';
import {
	buildPublicKeyPassphraseDict,
	sortKeysAscending,
	validateKeysSignatures,
	validateSignature,
} from './utils';
import { BaseTransactionInput, AccountAsset } from './types';

export const multisigRegAssetSchema = {
	$id: 'lisk/multisignature-registration-transaction',
	type: 'object',
	required: ['numberOfSignatures', 'optionalKeys', 'mandatoryKeys'],
	properties: {
		numberOfSignatures: {
			dataType: 'uint32',
			fieldNumber: 1,
			minimum: 1,
			maximum: 64,
		},
		mandatoryKeys: {
			type: 'array',
			items: {
				dataType: 'bytes',
			},
			fieldNumber: 2,
			minItems: 0,
			maxItems: 64,
			minLength: 32,
			maxLength: 32,
		},
		optionalKeys: {
			type: 'array',
			items: {
				dataType: 'bytes',
			},
			fieldNumber: 3,
			minItems: 0,
			maxItems: 64,
			minLength: 32,
			maxLength: 32,
		},
	},
};

const setMemberAccounts = async (
	store: StateStore,
	membersPublicKeys: Array<Readonly<Buffer>>,
): Promise<void> => {
	for (const memberPublicKey of membersPublicKeys) {
		const address = getAddressFromPublicKey(memberPublicKey as Buffer);
		// Key might not exists in the blockchain yet so we fetch or default
		const memberAccount = await store.account.getOrDefault(address);
		store.account.set(memberAccount.address, memberAccount);
	}
};

const hasDuplicateKey = (keys: Buffer[]): boolean => {
	const temp: { [key: string]: boolean | undefined } = {};
	for (const key of keys) {
		if (temp[key.toString('base64')]) {
			return true;
		}
		temp[key.toString('base64')] = true;
	}
	return false;
};

export interface MultiSignatureAsset {
	mandatoryKeys: Array<Readonly<Buffer>>;
	optionalKeys: Array<Readonly<Buffer>>;
	readonly numberOfSignatures: number;
}

export class MultisignatureTransaction extends BaseTransaction {
	public static TYPE = 12;
	public static ASSET_SCHEMA = multisigRegAssetSchema;
	public readonly asset: MultiSignatureAsset;
	private readonly MAX_KEYS_COUNT = 64;

	public constructor(transaction: BaseTransactionInput<MultiSignatureAsset>) {
		super(transaction);

		this.asset = transaction.asset;
	}

	// Verifies multisig signatures as per LIP-0017
	// eslint-disable-next-line @typescript-eslint/require-await
	public async verifySignatures(
		store: StateStore,
	): Promise<TransactionResponse> {
		const { networkIdentifier } = store.chain;
		const transactionBytes = this.getSigningBytes();
		const transactionWithNetworkIdentifierBytes = Buffer.concat([
			networkIdentifier,
			transactionBytes,
		]);

		const { mandatoryKeys, optionalKeys } = this.asset;

		// For multisig registration we need all signatures to be present
		if (
			mandatoryKeys.length + optionalKeys.length + 1 !==
			this.signatures.length
		) {
			return createResponse(this.id, [
				new TransactionError('There are missing signatures'),
			]);
		}

		// Check if empty signatures are present
		if (!this.signatures.every(signature => signature.length > 0)) {
			return createResponse(this.id, [
				new TransactionError(
					'A signature is required for each registered key.',
				),
			]);
		}

		// Verify first signature is from senderPublicKey
		const { valid, error } = validateSignature(
			this.senderPublicKey,
			this.signatures[0] as Buffer,
			transactionWithNetworkIdentifierBytes,
		);

		if (!valid) {
			return createResponse(this.id, [error as TransactionError]);
		}

		// Verify each mandatory key signed in order
		const mandatorySignaturesErrors = validateKeysSignatures(
			mandatoryKeys,
			this.signatures.slice(1, mandatoryKeys.length + 1),
			transactionWithNetworkIdentifierBytes,
		);

		if (mandatorySignaturesErrors.length) {
			return createResponse(this.id, mandatorySignaturesErrors);
		}
		// Verify each optional key signed in order
		const optionalSignaturesErrors = validateKeysSignatures(
			optionalKeys,
			this.signatures.slice(mandatoryKeys.length + 1),
			transactionWithNetworkIdentifierBytes,
		);
		if (optionalSignaturesErrors.length) {
			return createResponse(this.id, optionalSignaturesErrors);
		}

		return createResponse(this.id, []);
	}

	public sign(
		networkIdentifier: Buffer,
		senderPassphrase: string,
		passphrases?: ReadonlyArray<string>,
		keys?: {
			readonly mandatoryKeys: Array<Readonly<Buffer>>;
			readonly optionalKeys: Array<Readonly<Buffer>>;
			readonly numberOfSignatures: number;
		},
	): void {
		// Sort the keys in the transaction
		sortKeysAscending(this.asset.mandatoryKeys);
		sortKeysAscending(this.asset.optionalKeys);

		// Sign with sender
		const { publicKey } = getAddressAndPublicKeyFromPassphrase(
			senderPassphrase,
		);

		if (!this.senderPublicKey.equals(publicKey)) {
			throw new Error(
				'Transaction senderPublicKey does not match public key from passphrase',
			);
		}

		this.senderPublicKey = publicKey;

		const transactionWithNetworkIdentifierBytes = Buffer.concat([
			networkIdentifier,
			this.getSigningBytes(),
		]);

		this.signatures.push(
			signData(transactionWithNetworkIdentifierBytes, senderPassphrase),
		);

		// Sign with members
		if (keys && passphrases) {
			const keysAndPassphrases = buildPublicKeyPassphraseDict([...passphrases]);
			// Make sure passed in keys are sorted
			sortKeysAscending(keys.mandatoryKeys);
			sortKeysAscending(keys.optionalKeys);
			// Sign with all keys
			for (const aKey of [...keys.mandatoryKeys, ...keys.optionalKeys]) {
				const senderPublicKeyStr = bufferToHex(aKey as Buffer);
				// eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
				if (keysAndPassphrases[senderPublicKeyStr]) {
					const { passphrase } = keysAndPassphrases[senderPublicKeyStr];
					this.signatures.push(
						signData(transactionWithNetworkIdentifierBytes, passphrase),
					);
				} else {
					// Push an empty signature if a passphrase is missing
					this.signatures.push(Buffer.from(''));
				}
			}
		}
		this._id = hash(this.getBytes());
	}

	protected validateAsset(): ReadonlyArray<TransactionError> {
		const errors = [];

		const { mandatoryKeys, optionalKeys, numberOfSignatures } = this.asset;

		if (hasDuplicateKey(mandatoryKeys as Buffer[])) {
			errors.push(
				new TransactionError(
					'MandatoryKeys contains duplicate public keys',
					this.id,
					'.asset.mandatoryKeys',
				),
			);
		}

		if (hasDuplicateKey(optionalKeys as Buffer[])) {
			errors.push(
				new TransactionError(
					'OptionalKeys contains duplicate public keys',
					this.id,
					'.asset.optionalKeys',
				),
			);
		}

		// Check if key count is less than number of required signatures
		if (mandatoryKeys.length + optionalKeys.length < numberOfSignatures) {
			errors.push(
				new TransactionError(
					'The numberOfSignatures is bigger than the count of Mandatory and Optional keys',
					this.id,
					'.asset.numberOfSignatures',
					this.asset.numberOfSignatures,
				),
			);
		}

		// Check if key count is less than 1
		if (
			mandatoryKeys.length + optionalKeys.length > this.MAX_KEYS_COUNT ||
			mandatoryKeys.length + optionalKeys.length <= 0
		) {
			errors.push(
				new TransactionError(
					'The count of Mandatory and Optional keys should be between 1 and 64',
					this.id,
					'.asset.optionalKeys .asset.mandatoryKeys',
					this.asset.numberOfSignatures,
				),
			);
		}

		// The numberOfSignatures needs to be equal or bigger than number of mandatoryKeys
		if (mandatoryKeys.length > numberOfSignatures) {
			errors.push(
				new TransactionError(
					'The numberOfSignatures needs to be equal or bigger than the number of Mandatory keys',
					this.id,
					'.asset.numberOfSignatures',
					this.asset.numberOfSignatures,
				),
			);
		}

		if (errors.length > 0) {
			return errors;
		}

		// Check if keys are repeated between mandatory and optional key sets
		const repeatedKeys = mandatoryKeys.filter(
			value =>
				optionalKeys.find(optional => optional.equals(value as Buffer)) !==
				undefined,
		);
		if (repeatedKeys.length > 0) {
			errors.push(
				new TransactionError(
					'Invalid combination of Mandatory and Optional keys',
					this.id,
					'.asset.mandatoryKeys, .asset.optionalKeys',
					repeatedKeys.join(', '),
				),
			);
		}

		if (errors.length > 0) {
			return errors;
		}

		// Check if the length of mandatory, optional and sender keys matches the length of signatures
		if (
			mandatoryKeys.length + optionalKeys.length + 1 !==
			this.signatures.length
		) {
			return [
				new TransactionError(
					'The number of mandatory, optional and sender keys should match the number of signatures',
					this.id,
				),
			];
		}

		// Check keys are sorted lexicographically
		const sortedMandatoryKeys = [...mandatoryKeys].sort((a, b) =>
			a.compare(b as Buffer),
		);
		const sortedOptionalKeys = [...optionalKeys].sort((a, b) =>
			a.compare(b as Buffer),
		);
		for (let i = 0; i < sortedMandatoryKeys.length; i += 1) {
			if (!mandatoryKeys[i].equals(sortedMandatoryKeys[i] as Buffer)) {
				errors.push(
					new TransactionError(
						'Mandatory keys should be sorted lexicographically',
						this.id,
						'.asset.mandatoryKeys',
						mandatoryKeys.join(', '),
					),
				);
				break;
			}
		}

		for (let i = 0; i < sortedOptionalKeys.length; i += 1) {
			if (!optionalKeys[i].equals(sortedOptionalKeys[i] as Buffer)) {
				errors.push(
					new TransactionError(
						'Optional keys should be sorted lexicographically',
						this.id,
						'.asset.optionalKeys',
						optionalKeys.join(', '),
					),
				);
				break;
			}
		}

		return errors;
	}

	protected async applyAsset(
		store: StateStore,
	): Promise<ReadonlyArray<TransactionError>> {
		const errors: TransactionError[] = [];
		const sender = await store.account.get<AccountAsset>(this.senderId);

		// Check if multisignatures already exists on account
		if (sender.keys.numberOfSignatures > 0) {
			errors.push(
				new TransactionError(
					'Register multisignature only allowed once per account.',
					this.id,
					'.signatures',
				),
			);
		}

		sender.keys = {
			numberOfSignatures: this.asset.numberOfSignatures,
			mandatoryKeys: this.asset.mandatoryKeys as Buffer[],
			optionalKeys: this.asset.optionalKeys as Buffer[],
		};

		store.account.set(sender.address, sender);

		// Cache all members public keys
		await setMemberAccounts(store, sender.keys.mandatoryKeys);
		await setMemberAccounts(store, sender.keys.optionalKeys);

		return errors;
	}
}
