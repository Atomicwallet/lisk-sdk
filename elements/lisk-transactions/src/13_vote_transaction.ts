/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */

import { BaseTransaction, StateStore } from './base_transaction';
import { MAX_INT64 } from './constants';
import { TransactionError } from './errors';
import { sortUnlocking } from './utils';
import { BaseTransactionInput, AccountAsset } from './types';

export interface Vote {
	readonly delegateAddress: Buffer;
	readonly amount: bigint;
}

export interface RawAssetVote {
	readonly delegateAddress: string;
	readonly amount: string;
}

export interface VoteAsset {
	readonly votes: ReadonlyArray<Vote>;
}

const voteAssetSchema = {
	$id: 'lisk/vote-transaction',
	type: 'object',
	required: ['votes'],
	properties: {
		votes: {
			type: 'array',
			items: {
				type: 'object',
				required: ['delegateAddress', 'amount'],
				properties: {
					delegateAddress: {
						dataType: 'bytes',
						fieldNumber: 1,
						minLength: 20,
						maxLength: 20,
					},
					amount: {
						dataType: 'sint64',
						fieldNumber: 2,
					},
				},
			},
			fieldNumber: 1,
		},
	},
};

const TEN_UNIT = BigInt(10) * BigInt(10) ** BigInt(8);
const MAX_VOTE = 10;
const MAX_UNLOCKING = 20;

export class VoteTransaction extends BaseTransaction {
	public static TYPE = 13;
	public static ASSET_SCHEMA = voteAssetSchema;
	public readonly asset: VoteAsset;

	public constructor(transaction: BaseTransactionInput<VoteAsset>) {
		super(transaction);

		this.asset = transaction.asset;
	}

	protected validateAsset(): ReadonlyArray<TransactionError> {
		const errors = [];
		let upvoteCount = 0;
		let downvoteCount = 0;
		const addressSet = new Set();
		for (const vote of this.asset.votes) {
			addressSet.add(vote.delegateAddress);
			if (vote.amount > BigInt(0)) {
				upvoteCount += 1;
			} else if (vote.amount < BigInt(0)) {
				downvoteCount += 1;
			} else {
				errors.push(
					new TransactionError(
						'Amount cannot be 0',
						this.id,
						'.asset.votes.amount',
					),
				);
			}
			if (vote.amount % TEN_UNIT !== BigInt(0)) {
				errors.push(
					new TransactionError(
						'Amount should be multiple of 10 * 10^8',
						this.id,
						'.asset.votes.amount',
						vote.amount.toString(),
					),
				);
			}
		}
		if (upvoteCount > MAX_VOTE) {
			errors.push(
				new TransactionError(
					'Upvote can only be casted upto 10',
					this.id,
					'.asset.votes',
					upvoteCount,
					'10',
				),
			);
		}
		if (downvoteCount > MAX_VOTE) {
			errors.push(
				new TransactionError(
					'Downvote can only be casted upto 10',
					this.id,
					'.asset.votes',
					upvoteCount,
					'10',
				),
			);
		}
		if (addressSet.size !== this.asset.votes.length) {
			errors.push(
				new TransactionError(
					'Delegate address must be unique',
					this.id,
					'.asset.votes.delegateAddress',
				),
			);
		}

		return errors;
	}

	protected async applyAsset(
		store: StateStore,
	): Promise<ReadonlyArray<TransactionError>> {
		// Only order should be change, so no need to copy object itself
		const assetCopy = [...this.asset.votes];
		// Sort by ascending amount
		assetCopy.sort((a, b) => {
			const diff = a.amount - b.amount;
			if (diff > BigInt(0)) {
				return 1;
			}
			if (diff < BigInt(0)) {
				return -1;
			}

			return 0;
		});
		const errors = [];

		for (const vote of assetCopy) {
			const sender = await store.account.get<AccountAsset>(this.senderId);
			const votedDelegate = await store.account.getOrDefault<AccountAsset>(
				vote.delegateAddress,
			);

			if (!votedDelegate.asset.delegate.username) {
				errors.push(
					new TransactionError(
						'Voted delegate is not registered',
						this.id,
						'.asset.votes.delegateAddress',
					),
				);
				// eslint-disable-next-line no-continue
				continue;
			}
			if (vote.amount < BigInt(0)) {
				const originalUpvoteIndex = sender.asset.sentVotes.findIndex(
					senderVote => senderVote.delegateAddress.equals(vote.delegateAddress),
				);
				if (originalUpvoteIndex < 0) {
					errors.push(
						new TransactionError(
							'Cannot cast downvote to delegate who is not upvoted',
							this.id,
							'.asset.votes.delegateAddress',
						),
					);
					// eslint-disable-next-line no-continue
					continue;
				}
				sender.asset.sentVotes[originalUpvoteIndex].amount += vote.amount;
				if (sender.asset.sentVotes[originalUpvoteIndex].amount < BigInt(0)) {
					errors.push(
						new TransactionError(
							'Cannot downvote more than upvoted',
							this.id,
							'.asset.votes.amount',
						),
					);
				}
				// Delete entry when amount becomes 0
				if (sender.asset.sentVotes[originalUpvoteIndex].amount === BigInt(0)) {
					sender.asset.sentVotes = sender.asset.sentVotes.filter(
						senderVote => senderVote.delegateAddress !== vote.delegateAddress,
					);
				}
				// Create unlocking object
				sender.asset.unlocking.push({
					delegateAddress: vote.delegateAddress,
					amount: BigInt(-1) * vote.amount,
					unvoteHeight: store.chain.lastBlockHeader.height + 1,
				});
				// Sort account.unlocking
				sortUnlocking(sender.asset.unlocking);

				// Unlocking object should not exceed maximum
				if (sender.asset.unlocking.length > MAX_UNLOCKING) {
					errors.push(
						new TransactionError(
							`Cannot downvote which exceeds account.unlocking to have more than ${MAX_UNLOCKING.toString()}`,
							this.id,
							'.asset.votes',
						),
					);
				}
			} else {
				// Upvote amount case
				const originalUpvoteIndex = sender.asset.sentVotes.findIndex(
					senderVote => senderVote.delegateAddress.equals(vote.delegateAddress),
				);
				const index =
					originalUpvoteIndex > -1
						? originalUpvoteIndex
						: sender.asset.sentVotes.length;
				const upvote =
					originalUpvoteIndex > -1
						? sender.asset.sentVotes[originalUpvoteIndex]
						: {
								delegateAddress: vote.delegateAddress,
								amount: BigInt(0),
						  };
				upvote.amount += vote.amount;
				// Special case for postgres because maximum is int64 for bigint in postgres
				if (upvote.amount > BigInt(MAX_INT64)) {
					errors.push(
						new TransactionError(
							'Cannot upvote which exceeds int64',
							this.id,
							'.asset.votes.amount',
						),
					);
				}
				// Balance is checked in the base transaction
				sender.balance -= vote.amount;
				sender.asset.sentVotes[index] = upvote;
				// Sort account.votes
				sender.asset.sentVotes.sort((a, b) =>
					a.delegateAddress.compare(b.delegateAddress),
				);
				if (sender.asset.sentVotes.length > MAX_VOTE) {
					errors.push(
						new TransactionError(
							`Account can only vote upto ${MAX_VOTE.toString()}`,
							this.id,
							'.asset.votes.amount',
						),
					);
				}
			}
			store.account.set(sender.address, sender);
			// In case of self-vote, sender needs to be set and re-fetched to reflect both account change
			const delegate = await store.account.get<AccountAsset>(
				vote.delegateAddress,
			);
			delegate.asset.delegate.totalVotesReceived += vote.amount;
			store.account.set(delegate.address, delegate);
		}

		return errors;
	}

	protected async undoAsset(
		store: StateStore,
	): Promise<ReadonlyArray<TransactionError>> {
		const assetCopy = [...this.asset.votes];
		// Sort by descending amount
		assetCopy.sort((a, b) => {
			const diff = b.amount - a.amount;
			if (diff > BigInt(0)) {
				return 1;
			}
			if (diff < BigInt(0)) {
				return -1;
			}

			return 0;
		});
		for (const vote of assetCopy) {
			const sender = await store.account.get<AccountAsset>(this.senderId);
			if (vote.amount < BigInt(0)) {
				const originalUpvoteIndex = sender.asset.sentVotes.findIndex(
					senderVote => senderVote.delegateAddress.equals(vote.delegateAddress),
				);
				const index =
					originalUpvoteIndex > -1
						? originalUpvoteIndex
						: sender.asset.sentVotes.length;
				// If upvote does not exist anymore, it needs to re-adde them
				const upvote =
					originalUpvoteIndex > -1
						? sender.asset.sentVotes[originalUpvoteIndex]
						: {
								delegateAddress: vote.delegateAddress,
								amount: BigInt(0),
						  };
				// Add back the vote
				upvote.amount += vote.amount * BigInt(-1);
				sender.asset.sentVotes[index] = upvote;

				// Remove unlocking object
				const unlockingIndex = sender.asset.unlocking.findIndex(
					unlock =>
						unlock.delegateAddress.equals(vote.delegateAddress) &&
						unlock.amount === vote.amount * BigInt(-1) &&
						unlock.unvoteHeight === store.chain.lastBlockHeader.height + 1,
				);
				if (unlockingIndex < 0) {
					throw new Error(
						'Invalid data. unlocking object should exist while undo',
					);
				}
				sender.asset.unlocking.splice(unlockingIndex, 1);

				// Sort votes in case of reading
				sender.asset.sentVotes.sort((a, b) =>
					a.delegateAddress.compare(b.delegateAddress),
				);
				// Sort account.unlocking
				sortUnlocking(sender.asset.unlocking);
			} else {
				const originalUpvoteIndex = sender.asset.sentVotes.findIndex(
					senderVote => senderVote.delegateAddress.equals(vote.delegateAddress),
				);
				if (originalUpvoteIndex < 0) {
					throw new Error('Invalid data. Upvote should exist while undo');
				}
				sender.asset.sentVotes[originalUpvoteIndex].amount -= vote.amount;
				if (sender.asset.sentVotes[originalUpvoteIndex].amount === BigInt(0)) {
					sender.asset.sentVotes.splice(originalUpvoteIndex, 1);
				}
				sender.balance += vote.amount;
				// Sort account.votes
				sender.asset.sentVotes.sort((a, b) =>
					a.delegateAddress.compare(b.delegateAddress),
				);
			}
			store.account.set(sender.address, sender);
			// In case of self-vote, sender needs to be set and re-fetched to reflect both account change
			const delegate = await store.account.get<AccountAsset>(
				vote.delegateAddress,
			);
			delegate.asset.delegate.totalVotesReceived += vote.amount * BigInt(-1);
			store.account.set(delegate.address, delegate);
		}

		return [];
	}
}
