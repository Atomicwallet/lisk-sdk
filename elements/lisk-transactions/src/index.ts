/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */

import { DelegateTransaction } from './10_delegate_transaction';
import { MultisignatureTransaction } from './12_multisignature_transaction';
import { VoteTransaction } from './13_vote_transaction';
import { UnlockTransaction } from './14_unlock_transaction';
import { ProofOfMisbehaviorTransaction } from './15_proof_of_misbehavior_transaction';
import { TransferTransaction } from './8_transfer_transaction';
import { BaseTransaction, StateStore } from './base_transaction';
import * as constants from './constants';
import { convertToAssetError, convertToTransactionError, TransactionError } from './errors';
import { createResponse, Status, TransactionResponse } from './response';
import { transactionInterface } from './schema';
import {
	getSigningBytes,
	validateTransactionSchema,
	signTransaction,
	signMultiSignatureTransaction,
} from './transaction_helper';
import { Account, TransactionJSON } from './types';
import {
	convertBeddowsToLSK,
	convertLSKToBeddows,
	prependMinusToPublicKeys,
	prependPlusToPublicKeys,
	validateSenderIdAndPublicKey,
	validateSignature,
	verifyMinRemainingBalance,
	verifyMultiSignatureTransaction,
} from './utils';

const exposedUtils = {
	convertBeddowsToLSK,
	convertLSKToBeddows,
	prependMinusToPublicKeys,
	prependPlusToPublicKeys,
	validateSignature,
	verifyMinRemainingBalance,
	validateSenderIdAndPublicKey,
	verifyMultiSignatureTransaction,
};

export {
	Account,
	BaseTransaction,
	getSigningBytes,
	StateStore,
	TransferTransaction,
	DelegateTransaction,
	VoteTransaction,
	MultisignatureTransaction,
	UnlockTransaction,
	createResponse,
	ProofOfMisbehaviorTransaction,
	signTransaction,
	signMultiSignatureTransaction,
	Status,
	TransactionResponse,
	TransactionJSON,
	TransactionError,
	validateTransactionSchema,
	transactionInterface,
	convertToAssetError,
	convertToTransactionError,
	constants,
	exposedUtils as utils,
};
