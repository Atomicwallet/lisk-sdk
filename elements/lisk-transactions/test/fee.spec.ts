/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */

import { getAddressAndPublicKeyFromPassphrase } from '@liskhq/lisk-cryptography';
import { getMinFee } from '../src';

const validAssetSchema = {
	$id: 'lisk/transfer-transaction',
	title: 'Transfer transaction asset',
	type: 'object',
	required: ['amount', 'recipientAddress', 'data'],
	properties: {
		amount: {
			dataType: 'uint64',
			fieldNumber: 1,
		},
		recipientAddress: {
			dataType: 'bytes',
			fieldNumber: 2,
			minLength: 20,
			maxLength: 20,
		},
		data: {
			dataType: 'string',
			fieldNumber: 3,
			minLength: 0,
			maxLength: 64,
		},
	},
};
const passphrase1 = 'trim elegant oven term access apple obtain error grain excite lawn neck';
const { publicKey: publicKey1 } = getAddressAndPublicKeyFromPassphrase(passphrase1);
const validTransaction = {
	moduleID: 2,
	assetID: 0,
	nonce: BigInt('1'),
	senderPublicKey: publicKey1,
	asset: {
		recipientAddress: Buffer.from('3a971fd02b4a07fc20aad1936d3cb1d263b96e0f', 'hex'),
		amount: BigInt('4008489300000000'),
		data: '',
	},
};

describe('fee', () => {
	describe('getMinFee', () => {
		it('should return minimum fee required to send to network', () => {
			const fee = getMinFee(validAssetSchema, validTransaction);
			expect(fee).toMatchSnapshot();
		});
	});
});
