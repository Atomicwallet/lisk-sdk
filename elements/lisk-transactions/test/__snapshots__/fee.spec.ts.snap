// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fee getMinFee should return minimum fee required to send to network 1`] = `143000n`;
